File: frontend/types/node.ts
export type TaskType = {
    id: string;
    orderIndex: number;
    name: string;
    color: string;
    nodeId: string;
    createdAt: Date;
    updatedAt: Date;
};

export type TaskStatus = {
    id: string;
    orderIndex: number | null;
    name: string;
    kind: "Custom" | "Completed";
    color: string;
    nodeId: string;
    deletable: boolean;
    draggable: boolean;
    createdAt: Date;
    updatedAt: Date;
};

export type NodeType = {
    id: string;
    rank: number;
    name: string;
    kind: "Folder" | "Project";
    parentNodeId: string | null;
    userId: string;
    children: NodeType[];
    taskTypes: TaskType[];
    taskStatuses: TaskStatus[];
    createdAt: Date;
    updatedAt: Date;
    collapsed?: boolean;
};

export type FlattenedItem = NodeType & {
    depth: number;
};

File: backend/src/main/java/com/example/taskmanager/app/entities/Node.java
package com.example.taskmanager.app.entities;

import com.example.taskmanager.app.entities.task.Task;
import com.example.taskmanager.user.entities.User;
import jakarta.persistence.*;
import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Entity
@EntityListeners(AuditingEntityListener.class)
@Table(name = "nodes")
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Getter
@Setter
public class Node {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;
    private Long rank;
    @Enumerated(EnumType.STRING)
    private NodeKindEnum kind;
    @Column(nullable = false)
    private String name;
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;
    @CreatedDate
    private LocalDateTime createdAt;
    @LastModifiedDate
    private LocalDateTime updatedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_node_id")
    private Node parentNode;

    @OneToMany(mappedBy = "parentNode", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Node> children;

    @OneToMany(mappedBy = "node", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<TaskType> taskTypes;

    @OneToMany(mappedBy = "node", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<TaskStatus> taskStatuses;

    @OneToMany(mappedBy = "project", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Task> tasks;
}
